quinn-proto/src/tests/mod.rs

```diff
@@ -2184,6 +2211,388 @@ fn packet_splitting_not_necessary_after_higher_mtu_discovered() {
     assert_eq!(pair.server.inbound.len(), 1);
 }
 
+#[test]
+fn single_ack_eliciting_packet_triggers_ack_after_delay() {
+    let _guard = subscribe();
+    let mut pair = Pair::default();
+    let (client_ch, _) = pair.connect();
+    pair.drive();
+
+    let stats_after_connect = pair.client_conn_mut(client_ch).stats();
+
+    let start = pair.time;
+    pair.client_conn_mut(client_ch).ping();
+    pair.drive_client(); // Send ping
+    pair.drive_server(); // Process ping
+    pair.drive_client(); // Give the client a chance to process an ack, so our assertion can fail
+
+    // Sanity check: the time hasn't advanced in the meantime)
+    assert_eq!(pair.time, start);
+
+    let stats_after_ping = pair.client_conn_mut(client_ch).stats();
+    assert_eq!(
+        stats_after_ping.frame_tx.ping - stats_after_connect.frame_tx.ping,
+        1
+    );
+    assert_eq!(
+        stats_after_ping.frame_rx.acks - stats_after_connect.frame_rx.acks,
+        0
+    );
+
+    pair.drive();
+    let stats_after_drive = pair.client_conn_mut(client_ch).stats();
+    assert_eq!(
+        stats_after_drive.frame_rx.acks - stats_after_ping.frame_rx.acks,
+        1
+    );
+
+    // The time is start + max_ack_delay
+    assert_eq!(pair.time, start + Duration::from_millis(25));
+
+    // Sanity check: no loss probe was sent, because the delayed ACK was received on time
+    assert_eq!(
+        stats_after_drive.frame_tx.ping - stats_after_connect.frame_tx.ping,
+        1
+    );
+}
+
+#[test]
+fn immediate_ack_triggers_ack() {
+    let _guard = subscribe();
+    let mut pair = Pair::default();
+    let (client_ch, _) = pair.connect();
+    pair.drive();
+
+    let acks_after_connect = pair.client_conn_mut(client_ch).stats().frame_rx.acks;
+
+    pair.client_conn_mut(client_ch).immediate_ack();
+    pair.drive_client(); // Send ping
+    pair.drive_server(); // Process ping
+    pair.drive_client(); // Give the client a chance to process the ack
+
+    let acks_after_ping = pair.client_conn_mut(client_ch).stats().frame_rx.acks;
+
+    assert_eq!(acks_after_ping - acks_after_connect, 1);
```

Comment:
Shouldn't we be checking that the *server* sent an immediate ACK, upon receiving the request from the client? Why is the client sending an ACK at all in this period?

Reply:
As far as I can see, the assertion is indeed checking that the server sent an immediate ACK and it was processed by the client (we are checking the `frame_rx` stats of the client).

Reply:
Must've misread, whoops.